### PR TITLE
NFC: Accept non-parsed apps in Mifare DESFire.

### LINKED ISSUE #2040: Mifare DESFire UID emulation not possible due to "Can not parse file"
### Describe the bug.

This is related to #1730, but I want to clarify that I do not expect the encrypted (and thus unreadable) data to be saved. 
What I do want is that the UID is saved and can be emulated later, by opening a "saved" card.

I only have one particular DESFire card, and it is possible to use a cloned version of that card in its system, by just emulating the UID. I know this works, because the Android app "NFC Card Emulator Pro (Root)" can do that.
Of course, only the UID can be cloned, no encrypted data, but the reader accepts it anyway.

Now, after reading this card, the data which is saved by the firmware  is 'a bit short', because of these problems
```
7618 [I][NfcWorker] Mifare DESFire detected
7832 [W][MifareDESFire] Bad DESFire GET_KEY_SETTINGS response
7842 [W][MifareDESFire] Bad DESFire GET_KEY_SETTINGS response
```
so the file ends after the `Application IDs:`
```
Filetype: Flipper NFC device
Version: 2
# Nfc device type can be UID, Mifare Ultralight, Mifare Classic
Device type: Mifare DESFire
# UID, ATQA and SAK are common for all formats
UID: 04 1B 2C EA 6B 3B 80
ATQA: 44 03
SAK: 20
# Mifare DESFire specific data
PICC Version: 04 01 01 01 00 16 05 04 01 01 01 04 16 05 04 1B 2C EA 6B 3B 80 BA 54 D3 9D 70 33 14
PICC Free Memory: 1760
PICC Change Key ID: 00
PICC Config Changeable: true
PICC Free Create Delete: true
PICC Free Directory List: true
PICC Key Changeable: true
PICC Max Keys: 01
PICC Key 0 Version: A2
Application Count: 2
Application IDs: 01 00 00 02 00 00
```

This causes the "Can not parse file" screen, and nothing can be done with this file.
So I suggest to accept such file (as it was written by the firmware previously) and only emulate the UID, no apps.
I will provide a simple PR for that.

### Reproduction

1. Switch on
2. Go to NFC, Read
3. Place a specific Mifare DESFire token under the device
4. Save the data to a `.nfc`file
5. Go to NFC, Saved
6. Select the file that was just written
7. "Can not parse file" is shown, nothing can be done with the file.

### Target

_No response_

### Logs

_No response_

### Anything else?

There is a follow-up bug - when trying to emulate the UID of that specific card immediately after the "Read", the external reader will _not_ accept the emulation; and this apparently continues until the Flipper is rebooted, also when using the data from the saved file (which I did after applying the mentioned PR).

### PR DESCRIPTION
**Pull request by mh-**
Fixes #2040

# What's new

Accept non-parsed apps in Mifare DESFire.

# Verification 

Load this file from "Saved" in the NFC app
```
Filetype: Flipper NFC device
Version: 2
# Nfc device type can be UID, Mifare Ultralight, Mifare Classic
Device type: Mifare DESFire
# UID, ATQA and SAK are common for all formats
UID: 04 1B 2C EA 6B 3B 80
ATQA: 44 03
SAK: 20
# Mifare DESFire specific data
PICC Version: 04 01 01 01 00 16 05 04 01 01 01 04 16 05 04 1B 2C EA 6B 3B 80 BA 54 D3 9D 70 33 14
PICC Free Memory: 1760
PICC Change Key ID: 00
PICC Config Changeable: true
PICC Free Create Delete: true
PICC Free Directory List: true
PICC Key Changeable: true
PICC Max Keys: 01
PICC Key 0 Version: A2
Application Count: 2
Application IDs: 01 00 00 02 00 00
```

Before the PR, you will get a "Can not parse" error; 
after the PR, it will be possible to the emulate the UID.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: lib/nfc/nfc_device.c [hunk 1]
```diff
@@ -627,7 +627,11 @@ bool nfc_device_load_mifare_df_data(FlipperFormat* file, NfcDevice* dev) {
                 *app_head = app;
                 app_head = &app->next;
             }
-            if(!parsed_apps) break;
+            if(!parsed_apps) {
```
Comment (gornekich) [inline lib/nfc/nfc_device.c:631]: Could you declare 
#define TAG "NfcDevice"
at the beginning and write logs just 
FURI_LOG_W(TAG, "Non-parsed apps found!");

And please delete commented code

**Review by gornekich — CHANGES_REQUESTED** [review]
Hello @mh- 
I tested your file and your fix. It works fine.
However I can't reproduce this issue with any DESFire card I have. Could you show what data flipper shows just after you read this card?

**Comment by mh-** [comment]
Hi @gornekich 
> However I can't reproduce this issue with any DESFire card I have. Could you show what data flipper shows just after you read this card?

These are screenshots:
IMG_2397
IMG_2398
IMG_2399

Relevant logs:
```
106068 [I][NfcWorker] Mifare DESFire detected
106280 [W][MifareDESFire] Bad DESFire GET_KEY_SETTINGS response
106290 [W][MifareDESFire] Bad DESFire GET_KEY_SETTINGS response
125401 [W][nfc_device.c] before if(app->key_settings)  <-- this one added by me
```


In the firmware, when saving the file,
```
static bool nfc_device_save_mifare_df_app(FlipperFormat* file, MifareDesfireApplication* app) {
    bool saved = false;
    FuriString *prefix, *key;
    prefix =
        furi_string_alloc_printf("Application %02x%02x%02x", app->id[0], app->id[1], app->id[2]);
    key = furi_string_alloc();
    uint8_t* tmp = NULL;

    do {
        FURI_LOG_W("nfc_device.c", "before if(app->key_settings)");
        if(app->key_settings) {
            FURI_LOG_W("nfc_device.c", "if(app->key_settings)");
```
`app->key_settings` is `NULL`

**Comment by mh-** [comment]
Also, this is what the NXP "TagInfo" Android app says about this particular card:
image
image
image
image

**Comment by LowSkillDeveloper** [comment]
I have the same problem  https://github.com/flipperdevices/flipperzero-firmware/issues/1904#issuecomment-1295952022 And after this fix, if I did not know about the existence of this bug, I would think that I was emulating the card in full, and not just its UID. 
Maybe it's worth to displaying a message on screen that only UID is emulated, without Application. At least until the possibility of full emulation is found.

**Comment by mh-** [comment]
@LowSkillDeveloper 
> after this fix, if I did not know about the existence of this bug, I would think that I was emulating the card in full, and not just its UID.

Are you sure? On my Flipper, after loading a DESFire file from "Saved", the menu explicitly offers "Emulate UID", and not "Emulate" like with other non-DESFire files.

**Comment by LowSkillDeveloper** [comment]
> Are you sure? On my Flipper, after loading a DESFire file from "Saved", the menu explicitly offers "Emulate UID", and not "Emulate" like with other non-DESFire files.

Ah, yes, you are right.
I just mean that it might be worth making a message that maybe the card may not perform all the functions or not work properly.
Perhaps the information in the log is enough, but not everyone reads them.

**Review by gornekich — APPROVED** [review]

**Review by skotopes — APPROVED** [review]

**Pull request merged**